crates/wasmtime/src/store/data.rs

```diff
@@ -121,10 +113,11 @@ where
 
     #[inline]
     fn index(&self, index: Stored<T>) -> &Self::Output {
-        assert!(
-            index.store_id() == self.id,
-            "object used with the wrong store"
-        );
+        index.assert_belongs_to(self.id);
+        // Note that if this is ever a performance bottleneck it should be safe
+        // to use unchecked indexing here because presence of a `Stored<T>` is
+        // proof of an item having been inserted into a store so after the store
```

Comment:
"... inserted into a store" --> add "and stores are never shrunk, so an index once valid is always valid", just for extra clarity?